static/js/narrow.js

```diff
@@ -269,17 +270,23 @@ export function activate(raw_operators, opts) {
     save_pre_narrow_offset_for_reload();
 
     let msg_data = new MessageListData({
-        filter: narrow_state.filter(),
+        filter,
         excludes_muted_topics,
     });
 
+    const mld_exists = mld_cache.has(filter);
+    if (mld_exists) {
+        msg_data = mld_cache.get(filter);
+    }
+
```

Comment:
I think it might make sense to split this into first a commit that just implements the `MessageListDataCache` data structure, and adds test for it, and then adds a commit to store entries, and then proceeds with logic for actually using the cache.  

Part of my reasoning is that I think the logic for when to use the cache may end up being somewhat complex, potentially involving comparing the range of messages we have available from different data sources, and considering questions like whether the contiguous region that we have actually borders the message we need to land at.  And that will all require some pretty thoughtful testing, and potentially a series of smaller changes (E.g. at first we might want to only use the MessageListData cache when we know it includes the target message for the narrow).